ebu_tt_live/node/deduplicator.py

```diff
@@ -0,0 +1,199 @@
+            if value is not None:
+                unique_val = ComparableElement(value)
+                # stores references of original <xml:id> to <my_hash>
+                _old_id_dict[value.id] = unique_val.my_hash
+                # stores references of <my_hash> to <tt:region>
+                hash_dict[unique_val.my_hash] = value
+
+    def AppendNewElements(self, element_list, element_to_append_to, _old_id_dict, \
+                          _new_id_dict, hash_dict):
+        for hash_val in hash_dict:
+            new_id = hash_dict.get(hash_val)
+
+            for old_element in element_list:
+                if old_element.id is new_id.id:
+                    element_to_append_to.append(new_id)
+
+            _new_id_dict[hash_val] = new_id.id
+
+def ReplaceNone(none_value):
+    if none_value is None:
+        return "|" # '|' is a non-legal character and this is used to prevent
+                   # collisions between similar attributes
+    else:
+        return none_value
+
+
+class ComparableElement:
+    def __init__(self, value):
+        self.value = value
+
+        attributeDict = value._AttributeMap.copy()
+        xml_id_attr = ExpandedName('http://www.w3.org/XML/1998/namespace', 'id')
+        attributeDict.pop(xml_id_attr, None)
+
+        sortedDict = sorted(attributeDict.items(), key=lambda t: t[0])
+
+        concatenatedStyleString = u''
+        for key,val in sortedDict:
+            styleValue = ReplaceNone(val.value(value))
+            concatenatedStyleString += str(styleValue)
+
+        for key,val in value.wildcardAttributeMap().items():
+            namespace = ReplaceNone(key.namespaceURI())
+            localName = key.localName()
+            wildcardValue = ReplaceNone(val)
+            concatenatedStyleString += namespace + localName + wildcardValue
+
+        self.my_hash = hash(concatenatedStyleString)
+
+    def __eq__(self, other):
+        return other and self.my_hash == other.my_hash
+
+    def __ne__(self, other):
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        return self.my_hash
+
+
+class ReplaceStylesAndRegions(RecursiveOperation):
+    _old_id_dict = None
+    _new_id_dict = None
+
+    def __init__(self, root_element, _old_id_dict, _new_id_dict):
+            super(ReplaceStylesAndRegions, self).__init__(
+                root_element
+            )
+
+            self._old_id_dict = _old_id_dict
+            self._new_id_dict = _new_id_dict
+
+    def _is_begin_timed(self, value):
+        pass
+
+    def _before_element(self, value, element=None, parent_binding=None, **kwargs):
+        pass
+
+    def _after_element(self, value, element=None, parent_binding=None, **kwargs):
+        pass
+
+    def _process_element(self, value, element=None, parent_binding=None, **kwargs):
+        # The latter part of this and the next test is to check that the instance
+        # is not a styling or layout element as these can also have style attributes
```

Comment:
Slight tweak to the useful comment here: styling and layout can't have style attributes but their style _elements_ present themselves in exactly the same way as style attributes on other elements, so we have to avoid getting confused by them.